### PR TITLE
Demo a sidecar mode within pod for java users

### PR DESCRIPTION
**Pull request by resouer**
Close: #12526
Related: #13389 

After digging out file sharing in sidecar mode for a few days, I realize we can use life-cycle handler to do  a `volumes-from` style work within a Pod, see the demo for details.

Although it's some kind of workaround, but it's simple & practical, and after promoted it in Alibaba & Alipay, this sidecar mode is surprisingly popular ...

### CODE REVIEW

**Comment by resouer** [comment]
Related too: #4926 cc @bgrant0607

**Comment by bgrant0607** [comment]
cc @timstclair @amygdala

File: examples/javaweb-tomcat-sidecar/README.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+### Overview
+
+This sidecar mode brings a new workflow for Java users:
+
+![](workflow.png?raw=true "Workflow")
+
+As you can see, user can create a `sample:v2` container as sidecar to "provide" war file to Tomcat by copying it to the shared `emptyDir` volume. And Pod will make sure the two containers compose a "atomic" scheduling unit, which is perfect for this case. Thus, your application version management will be totally seperated from web server management.
+
+For example, if you gonna change the configurations of your Tomcat:
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: s/gonna/going to/

File: examples/javaweb-tomcat-sidecar/README.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/prereqs.md
+
+### Overview
+
+This sidecar mode brings a new workflow for Java users:
+
+![](workflow.png?raw=true "Workflow")
+
+As you can see, user can create a `sample:v2` container as sidecar to "provide" war file to Tomcat by copying it to the shared `emptyDir` volume. And Pod will make sure the two containers compose a "atomic" scheduling unit, which is perfect for this case. Thus, your application version management will be totally seperated from web server management.
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: "... an atomic scheduling ..."

File: examples/javaweb-tomcat-sidecar/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+Done! The new Tomcat image **will not** mess up with your `sample.war` file.
+
+You can also upgrade your app to new version seperately, without creating a new "Tomcat plus app" image (and this image will become huge as unionfs will keep all the old war files in its layers).
+
+**Why don't put my `sample.war` in a host dir and mount it to tomcat container?**
+
+You have to **manage the volumes** in this case, for example, when you restart or scale the Pod on another Node, your contents is not ready on that host.
+
+Generally, we have to set up a distributed file system (NFS at least) volume to solve this (if we do not have GCE PD volume). But seriously, it's a overkill.
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: s/it's a overkill/it's overkill/

Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: Actually, how about: "But this is generally unnecessary." instead?

File: examples/javaweb-tomcat-sidecar/README.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+### How To
+
+In Kubernetes a [_Pod_](../../docs/user-guide/pods.md) is the smallest deployable unit that can be created, scheduled, and managed. Its a collocated group of containers that share an IP and storage volume.
+
+Here is the config [javaweb.yaml](javaweb.yaml) for Java Web pod:
+
+NOTE: you should define `war` contaienr **first** as it is the "provider".
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: s/contaienr/container/

File: examples/javaweb-tomcat-sidecar/README.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+And the content inf `mv.sh` is:
+
+```sh
+cp /sample.war /app
+tail -f /etc/hosts
+```
+
+#### Explaination
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: Explanation

File: examples/javaweb-tomcat-sidecar/README.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+And the `resouer/sample:v2` Dockerfile is quite simple:
+
+```
+FROM busybox:latest
+ADD sample.war sample.war
+CMD "tail" "-f" "/etc/hosts"
+```
+
+#### Explaination
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: As above.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+```
+docker exec -it <tomcat_container_id> /bin/bash
+# do your change, and then commit it to a new image
+docker commit <tomcat_container_id> mytomcat:7.0-dev
+```
+
+Done! The new Tomcat image **will not** mess up with your `sample.war` file.
+
+You can also upgrade your app to new version seperately, without creating a new "Tomcat plus app" image (and this image will become huge as unionfs will keep all the old war files in its layers).
+
```
Comment (brendandburns) [inline examples/javaweb-tomcat-sidecar/README.md]: You might also want to mention that you can re-use your tomcat image with lots of different `war` container images for lots of different apps without having to build lots of different images.

Also this means that rolling out a new Tomcat to patch security or whatever else, doesn't require rebuilding N different images.

**Comment by brendandburns** [comment]
Generally LGTM.  Some spelling/typo/language style nits.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    name: tomcat
+    command: ["sh","-c","/root/apache-tomcat-7.0.42-v2/bin/start.sh"]
+    volumeMounts:
+    - mountPath: /root/apache-tomcat-7.0.42-v2/webapps
+      name: app-volume
+    ports:
+    - containerPort: 8080
+      hostPort: 8001
+  volumes:
+  - name: app-volume
```
Comment (timstclair) [inline examples/javaweb-tomcat-sidecar/README.md]: I believe implicit emptyDir volumes are deprecated, and you should specify the source here (same below)

File: examples/javaweb-tomcat-sidecar/README.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+The only magic here is the `resouer/sample:v1` image:
+
+```
+FROM busybox:latest
+ADD sample.war sample.war
+CMD "sh" "mv.sh"
+```
+
+And the content inf `mv.sh` is:
```
Comment (timstclair) [inline examples/javaweb-tomcat-sidecar/README.md]: s/content inf/contents of/

File: examples/javaweb-tomcat-sidecar/README.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+```sh
+cp /sample.war /app
+tail -f /etc/hosts
```
Comment (timstclair) [inline examples/javaweb-tomcat-sidecar/README.md]: If `sleep infinity` doesn't work here, I think `tail -f /dev/null` is more clear.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+# do your change, and then commit it to a new image
+docker commit <tomcat_container_id> mytomcat:7.0-dev
+```
+
+Done! The new Tomcat image **will not** mess up with your `sample.war` file.
+
+You can also upgrade your app to new version seperately, without creating a new "Tomcat plus app" image (and this image will become huge as unionfs will keep all the old war files in its layers).
+
+**Why don't put my `sample.war` in a host dir and mount it to tomcat container?**
+
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/Why don't/Why not

File: examples/javaweb-tomcat-sidecar/README.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+![](workflow.png?raw=true "Workflow")
+
+As you can see, user can create a `sample:v2` container as sidecar to "provide" war file to Tomcat by copying it to the shared `emptyDir` volume. And Pod will make sure the two containers compose a "atomic" scheduling unit, which is perfect for this case. Thus, your application version management will be totally seperated from web server management.
+
+For example, if you gonna change the configurations of your Tomcat:
+
+```
+docker exec -it <tomcat_container_id> /bin/bash
+# do your change, and then commit it to a new image
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/do your change/make some change 
May be better?

File: examples/javaweb-tomcat-sidecar/README.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+Done! The new Tomcat image **will not** mess up with your `sample.war` file.
+
+You can also upgrade your app to new version seperately, without creating a new "Tomcat plus app" image (and this image will become huge as unionfs will keep all the old war files in its layers).
+
+**Why don't put my `sample.war` in a host dir and mount it to tomcat container?**
+
+You have to **manage the volumes** in this case, for example, when you restart or scale the Pod on another Node, your contents is not ready on that host.
+
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/Pod/pod
s/Node/node
s/your contents is not ready/your file contents may not be ready

File: examples/javaweb-tomcat-sidecar/README.md (hunk 12 above)
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: should add a `$` before `docker` cmd.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+As you can see, user can create a `sample:v2` container as sidecar to "provide" war file to Tomcat by copying it to the shared `emptyDir` volume. And Pod will make sure the two containers compose a "atomic" scheduling unit, which is perfect for this case. Thus, your application version management will be totally seperated from web server management.
+
+For example, if you gonna change the configurations of your Tomcat:
+
+```
+docker exec -it <tomcat_container_id> /bin/bash
+# do your change, and then commit it to a new image
+docker commit <tomcat_container_id> mytomcat:7.0-dev
+```
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: Also here.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+You can also upgrade your app to new version seperately, without creating a new "Tomcat plus app" image (and this image will become huge as unionfs will keep all the old war files in its layers).
+
+**Why don't put my `sample.war` in a host dir and mount it to tomcat container?**
+
+You have to **manage the volumes** in this case, for example, when you restart or scale the Pod on another Node, your contents is not ready on that host.
+
+Generally, we have to set up a distributed file system (NFS at least) volume to solve this (if we do not have GCE PD volume). But seriously, it's a overkill.
+
+### How To
+
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/How to/How To Set it Up

File: examples/javaweb-tomcat-sidecar/README.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+In Kubernetes a [_Pod_](../../docs/user-guide/pods.md) is the smallest deployable unit that can be created, scheduled, and managed. Its a collocated group of containers that share an IP and storage volume.
+
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/Its/It's

File: examples/javaweb-tomcat-sidecar/README.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+```sh
+cp /sample.war /app
+tail -f /etc/hosts
+```
+
+#### Explaination
+
+1. 'war' container only contains the `war` file of your app
+2. 'war' container's CMD try to copy `sample.war` to the `emptyDir` volume path
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/try/tries

File: examples/javaweb-tomcat-sidecar/README.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+3. The last line of `tailf -f` is just used to hold the container, as RC does not support one-off task
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/RC/replication controller
I guess RC is not clear enough.

File: examples/javaweb-tomcat-sidecar/README.md [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+4. 'tomcat' container will load the `sample.war` from volume path
+
+What's more, if you don't want to add a build in `mv.sh` script in the `war` container, you can use Pod lifecycle handler to do the copy work, here's a example [javaweb-2.yaml](javaweb-2.yaml):
+
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/add a build in/enclose a built-in

File: examples/javaweb-tomcat-sidecar/README.md [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+```
+FROM busybox:latest
+ADD sample.war sample.war
+CMD "tail" "-f" "/etc/hosts"
+```
+
+#### Explaination
+
+1. 'war' container only contains the `war` file of your app
+2. 'war' container's CMD use `tail` to hold the container, nothing else
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: s/use/uses
"nothing else" is pretty odd, to be honest:)

File: examples/javaweb-tomcat-sidecar/README.md [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+
+Done! Now your `war` container contains nothing except `sample.war`, clean enough.
+
+### Test It Out
+
+Create the Java web pod:
+
+```sh
+kubectl create -f examples/javaweb-tomcat-sidecar/javaweb-2.yaml
+```
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: `$`

File: examples/javaweb-tomcat-sidecar/README.md [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+Check status of the pod:
+
+```sh
+kubectl get -w po
+NAME        READY     STATUS    RESTARTS   AGE
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: `$`

File: examples/javaweb-tomcat-sidecar/README.md [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+Create the Java web pod:
+
+```sh
+kubectl create -f examples/javaweb-tomcat-sidecar/javaweb-2.yaml
+```
+
+Check status of the pod:
+
+```sh
+kubectl get -w po
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: `$`

File: examples/javaweb-tomcat-sidecar/README.md [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+NAME        READY     STATUS    RESTARTS   AGE
+javaweb-2   2/2       Running   0         7s
+```
+
+Wait for the status to `2/2` and `Running`. Then you can visit "Hello, World" on `http://localhost:8001/sample/index.html`
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: `visit “Hello, World"` is pretty weird

File: examples/javaweb-tomcat-sidecar/README.md [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+You can also test `javaweb.yaml` in the same way.
+
+### Delete Resources
+
+All resources created in this application can be deleted:
+
+```sh
+kubectl delete -f examples/javaweb-tomcat-sidecar/javaweb-2.yaml
+```
```
Comment (dalanlan) [inline examples/javaweb-tomcat-sidecar/README.md]: `$`

**Comment by resouer** [comment]
Thanks for review, just updated. @brendandburns @dalanlan

File: examples/javaweb-tomcat-sidecar/README.md [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+
+Create the Java web pod:
+
+```sh
+$ kubectl create -f examples/javaweb-tomcat-sidecar/javaweb-2.yaml
+```
+
+Check status of the pod:
+
+```sh
```
Comment (bgrant0607) [inline examples/javaweb-tomcat-sidecar/README.md]: Should be ```console

(applies to all)

**Comment by brendandburns** [comment]
LGTM, but you need to update the examples test.

Please run `./hack/test-go.sh examples`

Error:
```
            examples_test.go:398: Could not get codec for &{{ } {      %!s(int64=0) 0001-01-01 00:00:00 +0000 UTC <nil> %!s(*int64=<nil>) map[] map[]} {%!s(int=0) map[] %!s(*api.PodTemplateSpec=<nil>) { %!s(*experimental.RollingUpdateDeployment=<nil>)} } {%!s(int=0) %!s(int=0)}}: unexpected kind: Deployment
panic: runtime error: invalid memory address or nil pointer dereference
/root/.gvm/gos/go1.3/src/pkg/runtime/panic.c:279 +0xf5
/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:416 +0x23f
/root/.gvm/gos/go1.3/src/pkg/runtime/panic.c:248 +0x18d
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/examples/examples_test.go:400 +0x8ba
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/examples/examples_test.go:152 +0x613
/root/.gvm/gos/go1.3/src/pkg/path/filepath/path.go:343 +0x9d
/root/.gvm/gos/go1.3/src/pkg/path/filepath/path.go:368 +0x60b
/root/.gvm/gos/go1.3/src/pkg/path/filepath/path.go:390 +0x10b
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/examples/examples_test.go:155 +0xb1
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/examples/examples_test.go:408 +0x8e05
/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:422 +0x110
/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:504 +0xb46
/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:505 +0xbb8
/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:435 +0xa3
k8s.io/kubernetes/examples/_test/_testmain.go:49 +0xdd
/root/.gvm/gos/go1.3/src/pkg/runtime/proc.c:1369 +0x89
/root/.gvm/gos/go1.3/src/pkg/runtime/proc.c:1385 +0x3b
/root/.gvm/gos/go1.3/src/pkg/runtime/mgc0.c:2644 +0xcf
/root/.gvm/gos/go1.3/src/pkg/runtime/proc.c:1445
/root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x96
/root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2d6
/root/.gvm/gos/go1.3/src/pkg/runtime/proc.c:1445
```

**Comment by resouer** [comment]
@brendandburns yes ... I also noticed that, but can not re-produce it:
```
$ ./hack/test-go.sh examples  
Running tests for APIVersion: v1,experimental/v1alpha1 with etcdPrefix: registry
+++ [0926 08:05:57] Running tests without code coverage
ok      k8s.io/kubernetes/examples      0.074s
Running tests for APIVersion: v1,experimental/v1alpha1 with etcdPrefix: kubernetes.io/registry
+++ [0926 08:06:02] Running tests without code coverage
ok      k8s.io/kubernetes/examples      0.060s
```
I already re-based & added a case in examples_test.go, but seems no clue about why shippable is still complaining ...  Any tips?

**Comment by resouer** [comment]
ping @brendandburns  could you please lead me out of here if you have time :)

**Comment by brendandburns** [comment]
LGTM, jenkins passes so I guess the examples test is ok.  Shippable looks like a flake, I kicked it.

**Comment by derekmahar** [comment]
What is sidecar mode?

**Comment by derekmahar** [comment]
> What is sidecar mode?

[Java Web Application with Tomcat and Sidercar Container](https://github.com/kubernetes/kubernetes/blob/release-1.4/examples/javaweb-tomcat-sidecar/README.md#java-web-application-with-tomcat-and-sidercar-container) is a good example of the sidecar concept.

**Pull request merged**